src/hotspot/share/opto/reachability.cpp

```diff
@@ -0,0 +1,555 @@
+  for (Node* ctrl = n; ctrl != nullptr; ctrl = ctrl->in(0)) {
+    assert(ctrl->is_CFG(), "");
+    if (visited.test_set(ctrl->_idx)) {
+      return;
+    } else {
+      if (ctrl->is_Region()) {
+        worklist.push(ctrl, 1);
+        return; // stop at merge points
+      } else if (ctrl->is_SafePoint() && is_interfering_sfpt_candidate(ctrl->as_SafePoint())) {
+        safepoints.push(ctrl);
+      }
+    }
+  }
+}
+
+// Enumerate all safepoints which are reachable from the RF to its referent through CFG.
+// Start at RF node and traverse CFG upwards until referent's control node is reached.
+static void enumerate_interfering_sfpts(ReachabilityFenceNode* rf, PhaseIdealLoop* phase, Node_List& safepoints) {
+  Node* referent = rf->referent();
+  Node* referent_ctrl = phase->get_ctrl(referent);
+  assert(phase->is_dominator(referent_ctrl, rf), "sanity");
+
+  VectorSet visited;
+  visited.set(referent_ctrl->_idx); // end point
+
+  Node_Stack stack(0);
+  linear_traversal(rf, stack, visited, safepoints); // start point
+  while (stack.is_nonempty()) {
+    Node* cur = stack.node();
+    uint  idx = stack.index();
+
+    assert(cur != nullptr, "");
+    assert(cur->is_Region(), "%s", NodeClassNames[cur->Opcode()]);
+    assert(phase->is_dominator(referent_ctrl, cur), "");
+    assert(idx > 0 && idx <= cur->req(), "%d %d", idx, cur->req());
+
+    if (idx < cur->req()) {
+      stack.set_index(idx + 1);
+      linear_traversal(cur->in(idx), stack, visited, safepoints);
+    } else {
+      stack.pop();
+    }
+  }
+}
+
+// Start offset for reachability info on a safepoint node.
+static uint rf_base_offset(SafePointNode* sfpt) {
+  return sfpt->jvms()->oopoff();
+}
+
+// Phase 2: migrate reachability info to safepoints.
+// All RFs are replaced with edges from corresponding referents to interfering safepoints.
+// Interfering safepoints are safepoint nodes which are reachable from the RF to its referent through CFG.
+bool PhaseIdealLoop::eliminate_reachability_fences() {
+  Compile::TracePhase tp(_t_reachability_eliminate);
+
+  assert(OptimizeReachabilityFences, "required");
+  assert(C->post_loop_opts_phase(), "required");
+  DEBUG_ONLY( int no_of_constant_rfs = 0; )
+
+  ResourceMark rm;
+  Unique_Node_List redundant_rfs;
+  Node_List worklist;
+  for (int i = 0; i < C->reachability_fences_count(); i++) {
+    ReachabilityFenceNode* rf = C->reachability_fence(i);
+    assert(!is_redundant_rf(rf, true /*rf_only*/), "missed");
+    if (PreserveReachabilityFencesOnConstants) {
+      const Type* referent_t = igvn().type(rf->referent());
+      assert(referent_t != TypePtr::NULL_PTR, "redundant rf");
+      bool is_constant_rf = referent_t->singleton();
+      if (is_constant_rf) {
+        DEBUG_ONLY( no_of_constant_rfs += 1; )
+        continue; // don't eliminate constant rfs
+      }
+    }
+    if (!is_redundant_rf(rf, false /*rf_only*/)) {
+      Node_List safepoints;
+      enumerate_interfering_sfpts(rf, this, safepoints);
```

Comment:
Could this explode if we have a lot of RF in the graph? Do we need a ResouceMark, or reuse the `safeponts` node list?

Imagine something like this:
```
referent
if (flag1) { something } else { something }
...
if (flag100) { something } else { something }
if (x1) { RF(referent); }
...
if (x100) { RF(referent); }
```
So we would call `enumerate_interfering_sfpts` 100x, and then traverse the graph with about 100-400 nodes each time. You can see how that grows quadratically. Maybe that's fine for runtime, but is it also ok for memory?

Reply:
And what if we find a lot of SafePoints for each RF? Do we end up attaching quadratically many referent edges over all?